tests/integration/test_sns.py

```diff
@@ -2985,3 +2910,284 @@ def test_message_structure_json_exc(self, sns_client, sns_create_topic, snapshot
+        sns_client,
+        sqs_client,
+        sqs_create_queue,
+        sns_create_topic,
+        sns_create_sqs_subscription,
+        snapshot,
+    ):
+        # https://docs.aws.amazon.com/sns/latest/dg/subscription-filter-policy-constraints.html
+        topic_arn = sns_create_topic()["TopicArn"]
+        queue_url = sqs_create_queue()
+        subscription = sns_create_sqs_subscription(topic_arn=topic_arn, queue_url=queue_url)
+        subscription_arn = subscription["SubscriptionArn"]
+
+        sns_client.set_subscription_attributes(
+            SubscriptionArn=subscription_arn,
+            AttributeName="FilterPolicyScope",
+            AttributeValue="MessageBody",
+        )
+
+        nested_filter_policy = {
+            "key_a": {
+                "key_b": {"key_c": ["value_one", "value_two", "value_three", "value_four"]},
+            },
+            "key_d": {"key_e": ["value_one", "value_two", "value_three"]},
+            "key_f": ["value_one", "value_two", "value_three"],
+        }
+        # The first array has four values in a three-level nested key, and the second has three values in a two-level
+        # nested key. The total combination is calculated as follows:
+        # 3 x 4 x 2 x 3 x 1 x 3 = 216
+        with pytest.raises(ClientError) as e:
+            sns_client.set_subscription_attributes(
+                SubscriptionArn=subscription_arn,
+                AttributeName="FilterPolicy",
+                AttributeValue=json.dumps(nested_filter_policy),
+            )
+        snapshot.match("sub-filter-policy-nested-error-too-many-combinations", e.value.response)
+
+        flat_filter_policy = {
+            "key_a": ["value_one"],
+            "key_b": ["value_two"],
+            "key_c": ["value_three"],
+            "key_d": ["value_four"],
+            "key_e": ["value_five"],
+            "key_f": ["value_six"],
+        }
+        # A filter policy can have a maximum of five attribute names. For a nested policy, only parent keys are counted.
+        with pytest.raises(ClientError) as e:
+            sns_client.set_subscription_attributes(
+                SubscriptionArn=subscription_arn,
+                AttributeName="FilterPolicy",
+                AttributeValue=json.dumps(flat_filter_policy),
+            )
+        snapshot.match("sub-filter-policy-max-attr-keys", e.value.response)
+
+        flat_filter_policy = {"key_a": "value_one"}
+        # Rules should be contained in a list
+        with pytest.raises(ClientError) as e:
+            sns_client.set_subscription_attributes(
+                SubscriptionArn=subscription_arn,
+                AttributeName="FilterPolicy",
+                AttributeValue=json.dumps(flat_filter_policy),
+            )
+        snapshot.match("sub-filter-policy-rule-no-list", e.value.response)
+        assert (
+            e.value.response["Error"]["Message"]
+            == 'Invalid parameter: FilterPolicy: "key_a" must be an object or an array'
```

Comment:
nit: single outer quotes that apparently were not caught by the linter? considering we have double outer quotes everywhere else. (But I guess it doesn't matter which ones are the outer ones)

Reply:
Yes, it's because they are inner double quotes for parity with the AWS error message, so I guess it's allowed by the linter. Thanks for the check!